test/runnable/test22227.d

```diff
@@ -0,0 +1,16 @@
+// REQUIRED_ARGS: -debug -O -release
+// https://issues.dlang.org/show_bug.cgi?id=22277
+
+bool secret = false;
+
+void free(immutable void* x) pure nothrow
+{
+    debug secret = true;
+}
+
+void main()
+{
+    free(null);
+    if (!secret)
```

Comment:
Why not simply `assert(secret)`?

Reply:
Because it's compiled with -release so that does nothing.

Reply:
So `assert(0)` is not removed in release mode. I find that surprising. Thanks.

Reply:
> I find that surprising and yet another language exception rule to remember.

It surprised me too when I first heard about it. The spec says this:

https://dlang.org/spec/expression.html#assert_expressions

> The implementation may handle the case of the first AssignExpression evaluating at compile time to false differently in that in release mode it may simply generate a HLT instruction or equivalent.

Reply:
What was the original motive behind adding this language exception?

Reply:
The motive behind removing asserts in release mode in the first place is to save time and space by not having a check and assert message. Since `assert(0)` can compile to a single HLT instruction in release mode, it's so cheap it's not worth removing entirely.

Reply:
`assert(0)` is also critical for memory safety. A function that ends in `assert(0)` can have a missing return statement relying on the fact that execution won't continue.

Reply:
Thanks.